Move SHOW and TERMINATE transactions to Ops Manual (#3034)
neo4j/docs-operations: modules/ROOT/pages/database-internals/list-and-terminate-transactions.adoc

| List transactions on the current serve

a|
[source, syntax, role="noheader"]
----
SHOW TRANSACTION[S] [transaction-id[,...]]
[YIELD { * \| field[, ...] } [ORDER BY field[, ...]] [SKIP n] [LIMIT n]]
[WHERE expression]
[RETURN field[, ...] [ORDER BY field[, ...]] [SKIP n] [LIMIT n]]
----

[NOTE]
====
The format of `transaction-id` is `<databaseName>-transaction-<id>`.
Transaction IDs must be supplied as one or more comma-separated quoted `STRING` values, or as an expression resolving to a `STRING` or a `LIST<STRING>`.
As of Neo4j 2026.06, using Cypher 25, any transaction IDs that resolve to `null` are ignored instead of causing a type error exception.
This is consistent with returning no results when the specified transaction ID does not exist.
====

|===

[NOTE]
====
When using the `RETURN` clause, the `YIELD` clause is mandatory and must not be omitted.
====

A user with the link:{neo4j-docs-base-uri}/operations-manual/current/authentication-authorization/database-administration/#access-control-database-administration-transaction[`SHOW TRANSACTION`] privilege can view currently executing transactions, but only to the extent allowed by their granted permissions. 

> All users can always view their own active transactions.

Reviewer: if we're anyway reordering things it might be nice to move this above the syntax section instead of part of it, it's not really syntax related 🤷

| transactionsCommitted
| Total number of inner transactions committed by the `CALL { ... } IN TRANSACTIONS` query, if this query is an outer transaction, otherwise `null`. For details, see 

| transactionsRolledBack
| Total number of inner transactions rolled back by the `CALL { ... } IN TRANSACTIONS` query, if this query is an outer transaction, otherwise `null`.
|===

For more information about `transactionsStarted`, `transactionsCommitted`, and `transactionsRolledBack` see the link:{neo4j-docs-base-uri}/cypher-manual/current/subqueries/subqueries-in-transactions/[Cypher Manual -> CALL subqueries in transactions].

[[listing-transactions-examples]]
=== Examples

.List all running transactions with default return columns
[source, cypher]
----
SHOW TRANSACTIONS
----

.List all running transactions with all return columns
[source, cypher]
----
SHOW TRANSACTIONS YIELD *
----

.List transactions by transactionId
[source, cypher]
----
SHOW TRANSACTIONS "neo4j-transaction-3"
----

.Filter `SHOW TRANSACTIONS` with `WHERE`
[source, cypher]
----
SHOW TRANSACTIONS YIELD database, currentQuery
WHERE currentQuery contains 'Mark'
----

> .Return renamed `SHOW TRANSACTIONS` columns

Reviewer: this is more about returning the time related columns as milliseconds/seconds/microseconds rather than timestamps/duration values and not about renamings

this is another case where it would be nice for the examples to have results as the reason for this example was to show how to get 'simpler' printed results from the durations (something like `PT0.038S` vs `1055`).

  currentQueryCpuTime.milliseconds AS currentQueryCpuTimeMillis,
  currentQueryWaitTime.microseconds AS currentQueryWaitTimeMicros,
  currentQueryIdleTime.seconds AS currentQueryIdleTimeSeconds
----

[[terminate-transactions]]
== Terminate transactions

The `TERMINATE TRANSACTIONS` command is used to terminate running transactions by their IDs.

[[terminate-transactions-syntax]]
=== Syntax

[options="header", width="100%", cols="1,5a"]
|===
| Action | Syntax

| Terminate transactions by ID on the current server

a|
[source, syntax, role="noheader"]
----
TERMINATE TRANSACTION[S] transaction_id[, ...]
[YIELD { * \| field[, ...] }
  [ORDER BY field[, ...]]
  [SKIP n]
  [LIMIT n]
  [WHERE expression]
  [RETURN field[, ...] [ORDER BY field[, ...]] [SKIP n] [LIMIT n]]
]
----

[NOTE]
====
The format of `transaction-id` is `<databaseName>-transaction-<id>`.
Transaction IDs must be supplied as one or more comma-separated quoted `STRING` values, or as an expression resolving to a `STRING` or a `LIST<STRING>`.
As of Neo4j 2026.06, using Cypher 25, any transaction IDs that resolve to `null` are ignored instead of causing a type error exception.

> This is consistent with returning no results when the specified transaction ID does not exist.

Reviewer: this is not the correct sentence for TERMINATE, it has different behaviour to non-existing ids than show.
```suggestion
As of Neo4j 2026.06, using Cypher 25, any transaction IDs resolves to `null` will no longer throw an exception on wrong type.
Instead, they will return a row similar to the row returned when the given transaction ID doesn't exist.
```

====
|===

[NOTE]
====
When using the `WHERE` or `RETURN` clauses, the `YIELD` clause is mandatory and must not be omitted.
====

A user with the link:{neo4j-docs-base-uri}/operations-manual/current/authentication-authorization/database-administration/#access-control-database-administration-transaction[`TERMINATE TRANSACTION`] privilege can terminate transactions, subject to the scope of their granted permissions.

> All users may terminate their own currently executing transactions.

Reviewer: similarly to he show case, might be nice to move this above the syntax section instead of as part of it

WHERE waitTime > duration({minutes: 30})
TERMINATE TRANSACTIONS transactionId
YIELD username, message
RETURN *
----
 
.Listing other transactions by the same user that were terminated
[source, cypher]
----
TERMINATE TRANSACTION 'neo4j-transaction-1', 'neo4j-transaction-2'
YIELD username AS terminatedUser
SHOW TRANSACTIONS
YIELD username AS showUser, transactionId AS txId, database, currentQuery, status
WHERE showUser = terminatedUser AND NOT status STARTS WITH 'Terminated'
RETURN txId, showUser AS user, database, currentQuery
----

.Listing other transactions by the same user as a given transaction
[source, cypher]
----
SHOW TRANSACTION 'neo4j-transaction-1'
YIELD username AS originalUser, transactionId AS originalTxId
SHOW TRANSACTIONS
YIELD username AS newUser, transactionId AS txId, database, currentQuery, status
WHERE newUser = originalUser AND NOT txId = originalTxId
RETURN txId, newUser AS user, database, currentQuery, status
----

[role=label--new-2026.05 label--cypher-25-only]
== Combine transaction commands with other Cypher commands

`SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS` can be combined with other Cypher commands to form a single query.

[NOTE]
====

> When combining a `SHOW` command with other clauses, the `YIELD` clause is mandatory and must not be omitted.

Reviewer: ```suggestion
When combining a `SHOW` or `TERMINATE` command with other clauses, the `YIELD` clause is mandatory and must not be omitted.
```

In addition, `YIELD *` is not permitted.
Instead, the `YIELD` clause needs to explicitly list the yielded columns.
The query must also end with a valid last clause (like a `RETURN` or an updating clause).
====

.Transaction commands combined with other Cypher commands
[source, cypher, role=test-result-skip]
----
WITH ['neo4j-transaction-1', 'neo4j-transaction-2', 'mydb-transaction-1'] AS idsToTerminate
TERMINATE TRANSACTION idsToTerminate
YIELD username AS terminatedUser
WITH DISTINCT terminatedUser, idsToTerminate // <1>
SHOW TRANSACTIONS
YIELD username AS showUser, transactionId AS txId, database, currentQuery
WHERE showUser = terminatedUser AND NOT txId IN idsToTerminate
RETURN txId, showUser AS user, database, currentQuery
----

> <1> `WITH` is used here to remove any duplicate users and transaction IDs and to bind those values to new variables, `terminatedUser` and `idsToTerminate`.

Reviewer: not sure I'd say we bind them to new variables 🤷 it's not like they are renamed in the WITH, they just keep existing from before

Reviewer: the sentence in my PR around "We can also save the transaction ids in a variable," has to do with the first line/with and not the one marked with a `// <1>`

:description: Information about the `SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS` commands.
:test-skip: true

> = List and terminate transactions

Reviewer 2: ```suggestion
= Manage transactions
```

> This page discusses how to use the `SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS` to list and terminate currently running transactions.

Reviewer 2: ```suggestion
Transactions can be managed using the Cypher commands `SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS`.
The `TERMINATE TRANSACTIONS` command can be combined with multiple `SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS` commands in the same query.
```

Reviewer: Not sure we need the second of your sentences Reneta, but I also think it sounds a bit odd to point out that TERMINATE TRANSACTIONS can be combined with others (when it's just generic) so might be better to keep it more generic:
```suggestion
Transactions can be managed using the Cypher commands `SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS`.
Multiple  `SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS` commands can be combined in the same query.
```
and if we add such a sentence we should probably also add the one about combining things in general that I added in my original PR
```suggestion
Transactions can be managed using the Cypher commands `SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS`.
Multiple  `SHOW TRANSACTIONS` and `TERMINATE TRANSACTIONS` commands can be combined in the same query.
As of Neo4j 2026.05, using Cypher 25, it is also possible to combine the transaction commands with general Cypher clauses.
```

A user with the link:{neo4j-docs-base-uri}/operations-manual/current/authentication-authorization/database-administration/#access-control-database-administration-transaction[`SHOW TRANSACTION`] privilege can view currently executing transactions, but only to the extent allowed by their granted permissions. 
All users can always view their own active transactions.

[[list-transactions-syntax]]
=== Syntax

For full details about the syntax descriptions, see xref:database-administration/syntax.adoc#administration-syntax-reading[Administration command syntax].

[options="header", width="100%", cols="1,5a"]
|===
| Action | Syntax

| List transactions on the current serve

a|
[source, syntax, role="noheader"]
----
SHOW TRANSACTION[S] [transaction-id[,...]]
[YIELD { * \| field[, ...] } [ORDER BY field[, ...]] [SKIP n] [LIMIT n]]
[WHERE expression]
[RETURN field[, ...] [ORDER BY field[, ...]] [SKIP n] [LIMIT n]]
----

[NOTE]
====
The format of `transaction-id` is `<databaseName>-transaction-<id>`.
Transaction IDs must be supplied as one or more comma-separated quoted `STRING` values, or as an expression resolving to a `STRING` or a `LIST<STRING>`.

> As of Neo4j 2026.06, using Cypher 25, any transaction IDs that resolve to `null` are ignored instead of causing a type error exception.

Reviewer: ```suggestion
As of Neo4j 2026.05, using Cypher 25, any transaction IDs that resolve to `null` are ignored instead of causing a type error exception.
```
